sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs

```diff
@@ -0,0 +1,196 @@
+            if (indexerProperty?.ExpressionBody != null)
+            {
+                // Arrowed property
+                return propertyReturnType
+                    .Union(indexerProperty.ExpressionBody
+                        .DescendantNodes()
+                        .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model)));
+            }
+
+            var basicProperty = property as PropertyDeclarationSyntax;
+            if (basicProperty?.ExpressionBody != null)
+            {
+                // Arrowed property
+                return propertyReturnType
+                    .Union(basicProperty.ExpressionBody
+                        .DescendantNodes()
+                        .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model)));
+            }
+
+            // Classic property OR event
+            return propertyReturnType
+                .Union(property.AccessorList.Accessors
+                    .SelectMany(accessor => accessor.DescendantNodes()
+                        .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model))));
+        }
+
+        private static IEnumerable<ITypeSymbol> CollectCoupledClasses(BaseMethodDeclarationSyntax method,
+            SemanticModel model)
+        {
+            return method.DescendantNodes()
+                .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model));
+        }
+
+        private static IEnumerable<ITypeSymbol> ExtractTypeSymbolsFromSyntaxNode(SyntaxNode node, SemanticModel model)
+        {
+            var parameter = node as ParameterSyntax;
+            if (parameter != null)
+            {
+                return parameter.Type != null
+                        ? new[] { model.GetSymbolInfo(parameter.Type).Symbol as ITypeSymbol }
+                        : new[] { (ITypeSymbol)null };
+            }
+
+            var variableDeclaration = node as VariableDeclarationSyntax;
+            if (variableDeclaration != null)
+            {
+                return new[] { model.GetSymbolInfo(variableDeclaration.Type).Symbol as ITypeSymbol };
+            }
+
+            var objectCreation = node as ObjectCreationExpressionSyntax;
+            if (objectCreation != null)
+            {
+                var objectCreationTypeInfo = model.GetTypeInfo(objectCreation);
+                return new[] { objectCreationTypeInfo.Type, objectCreationTypeInfo.ConvertedType };
+            }
+
+            var identifierName = node as IdentifierNameSyntax;
+            return identifierName != null
+                ? new[] { model.GetSymbolInfo(identifierName).Symbol as ITypeSymbol }
+                : new[] { (ITypeSymbol)null };
+        }
+
+        private static int FilterAndCountCoupling(IEnumerable<ITypeSymbol> types)
+        {
+            return types.Distinct()
+                .WhereNotNull()
+                .Where(type => type.TypeKind != TypeKind.Enum &&
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Drop 'Where' and move the condition into the 'Count'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2971)